Landing page updates

AndrewDiMola opened the pull request:
- Added sections for Async and XHP. 
- Removed Type Annotations and Lambdas section.
- Incorporated links to docs throughout.
- Added tagline.
- Added section for collaboration/research/usage.

AndrewDiMola committed:
Landing page updates

- Added sections for Async and XHP. - Removed Type Annotations and Lambdas section.
- Incorporpated links to docs throughout.
- Added tagline.
- Added section for collaboration/research/usage.

AndrewDiMola commented:
<img width="1670" alt="Screen Shot 2021-04-23 at 2 40 46 PM" src="https://user-images.githubusercontent.com/5179225/115916162-6be68b80-a442-11eb-85d6-6ec8fe2624f1.png">
<img width="1678" alt="Screen Shot 2021-04-23 at 2 40 56 PM" src="https://user-images.githubusercontent.com/5179225/115916164-6c7f2200-a442-11eb-84ca-8dd22a024fbd.png">
<img width="1672" alt="Screen Shot 2021-04-23 at 2 41 04 PM" src="https://user-images.githubusercontent.com/5179225/115916169-6d17b880-a442-11eb-9118-28142e867915.png">

AndrewDiMola commented:
@fredemmott @Wilfred @jjergus

fredemmott commented on website/pages/en/index.js:
````diff
@@ -14,16 +14,19 @@ const siteConfig = require(process.cwd() + "/siteConfig.js");
 const pre = "```";
-const typeAnnotationExample= `${pre}hack
+const asyncExample= `${pre}hack
 <?hh
````
remove - discouraged in new public code

AndrewDiMola replied:
can you clarify what/which to remove and what's discouraged? thanks!

fredemmott replied:
the `<?hh` line is only needed in `.php` files, and is invalid in `.hack` and `.hackpartial` files; while FB still mostly uses `.php` files, `.hack` are preferred externally

fredemmott commented on website/pages/en/index.js:
```diff
@@ -138,17 +141,20 @@ class Index extends React.Component {
                   <div>
                     <MarkdownBlock>
-                      Type annotations allow for code to be explicitly typed on parameters, class member variables and return values.
+                      **Async is not multithreading**. HHVM executes a program's code in one main request thread.
```
This is important, but here it's probably to swap these around to describe what it is, before what it isn't

fredemmott commented on website/siteConfig.js:
```diff
@@ -10,8 +10,8 @@
   title: "Hack" /* title for your website */,
-  /* tagline:  "foo" , */
-  disableTitleTagline: true,
+  tagline: "Moving fast, making changes daily, and releasing often",
```
Not convinced by this - 'making changes daily' is in 'implementation details' territory, and overall this feels more like a description of project management than the language itself

AndrewDiMola replied:
I'll just remove it entirely and switch the flag back for now.

jjergus replied:
I think we could add this as a one more section similar to "collaboration, research and usage", or maybe just a bullet point in that section. Something like

> ## Active Development
> HHVM and the Hack language are in active development. We are moving fast, making changes daily and releasing often. If you notice a regression in the typechecker or the runtime, please [open issues](https://github.com/facebook/hhvm/issues/new) when you find them.

AndrewDiMola replied:
Sure, I can add it as an introduction <p>

AndrewDiMola committed:
Update to Pull/28

- Removed typeahead from examples
- Removed tagline
- Added introduction paragraph to Collaborations, Research, and Usage
- Refactored Generics section

Wilfred commented on website/pages/en/index.js:
```diff
@@ -14,41 +14,41 @@ const siteConfig = require(process.cwd() + "/siteConfig.js");
-    $y = $x + 1;
-    return $y;
+const asyncExample= `${pre}hack
+  async function helloWorld(): Awaitable<string> {
```
We should probably follow Hack naming conventions for top-level functions :)

```suggestion
  async function hello_world(): Awaitable<string> {
```

AndrewDiMola replied:
🤭

Wilfred commented on website/pages/en/index.js:
```diff
@@ -14,41 +14,41 @@ const siteConfig = require(process.cwd() + "/siteConfig.js");
+const XHPExample = `${pre}hack
+  // Traditional: HTML elements require quotation marks
+  $user_name = 'Fred';
+  echo "<tt>Hello <strong>$user_name</strong></tt>";
```
XHP guarantees that your tags are well-formed. Perhaps show how it's easy to screw up with strings?
```suggestion
  echo "<tt>Hello <strong>$user_name</tt></strong>";
```

XHP also protects you from XSS vulnerabilities, but that's a little more verbose :)

Wilfred commented on website/pages/en/index.js:
```diff
@@ -138,17 +138,20 @@ class Index extends React.Component {
+                      **[Asynchronous operations](https://docs.hhvm.com/hack/asynchronous-operations/introduction)** allow _cooperative multi-tasking_. Code that utilizes the Async Infra can hide I/O latency and data fetching, minimizing the downtime a program has to be stalled for I/O or similar operations.
+                    </MarkdownBlock>
+                    <MarkdownBlock>
+                      **Async is not multithreading**. HHVM executes a program's code in one main request thread.
```
I don't think limitation details like this belong on the home page. We've already said it's cooperative multi-tasking.

Wilfred commented on website/pages/en/index.js:
```diff
@@ -164,7 +167,10 @@ class Index extends React.Component {
+                      **[Generics](https://docs.hhvm.com/hack/generics/some-basics)** allow classes and methods to be parameterized to any set of types, all while preserving type safety, and supports both **_[generic covariance and contravariance](https://docs.hhvm.com/hack/generics/variance)_** on a type parameter.
+                    </MarkdownBlock>
+                    <MarkdownBlock>
+                      You can also create **_[Reified Generics](https://docs.hhvm.com/hack/generics/reified-generics)_**—Generics with type information accessible at runtime.
```
Generics is a really big and powerful feature in Hack. Rather than enumerating a few things you can do with generics, perhaps we should say something more general?

> Hack supports a rich set of generic types, including type parameters, constraints, associated type constants and even reification.

Wilfred commented on website/pages/en/index.js:
```diff
@@ -184,17 +190,45 @@ class Index extends React.Component {
+                      * **[Jump-Start](https://engineering.fb.com/2021/03/03/developer-tools/hhvm-jump-start/)**: Jump-Start improves the performance of virtual machines at scale and has successfully been implemented in the HipHop Virtual Machine (HHVM), which powers not only Facebook.com but also many other sites across the web.
+                    </MarkdownBlock>
+                    <MarkdownBlock>
+                      * **[Zoncolan](https://engineering.fb.com/2019/08/15/security/zoncolan/)**: Facebook’s web codebase currently contains more than 100 million lines of Hack code, and changes thousands of times per day. With Zoncolan and its static analysis capabilities, security engineers scale their work by automatically examining Hack code and proactively detecting potentially dangerous sec ...
```
Zoncolan is really cool, but it's not available to OSS users. Could we describe this as "Recent research" or something, to focus that it's academic output in this section?

AndrewDiMola committed:
Update website/pages/en/index.js

Co-authored-by: Wilfred Hughes <me@wilfred.me.uk>

AndrewDiMola committed:
Update website/pages/en/index.js

Co-authored-by: Wilfred Hughes <me@wilfred.me.uk>

AndrewDiMola committed:
Suggestions #28

- Reworked XHP and Generics copy
- Fixed samples to adhere to best practices
- Clarified articles are on recent academic research

AndrewDiMola commented:
Sorry for the delay here -- all feedback (hopefully) incorporated in the last set of commits.

Wilfred commented on website/pages/en/index.js:
```diff
@@ -138,17 +138,17 @@ class Index extends React.Component {
                   <div>
                     <MarkdownBlock>
-                      Type annotations allow for code to be explicitly typed on parameters, class member variables and return values.
+                      **[Asynchronous operations](https://docs.hhvm.com/hack/asynchronous-operations/introduction)** allow _cooperative multi-tasking_. Code that utilizes the Async Infra can hide I/O latency and data fetching, minimizing the downtime a program has to be stalled for I/O or similar operations.
```
I might avoid "downtime" as async doesn't affect robustness. It just makes your program faster.

```suggestion
                      **[Asynchronous operations](https://docs.hhvm.com/hack/asynchronous-operations/introduction)** allow _cooperative multi-tasking_. Async functions can combine I/O requests, making your code faster.
```

Wilfred commented on website/pages/en/index.js:
```diff
@@ -14,41 +14,41 @@ const siteConfig = require(process.cwd() + "/siteConfig.js");
-    $y = $x + 1;
-    return $y;
+const asyncExample= `${pre}hack
+  async function hello_world(): Awaitable<string> {
+    return "Hello World"; // simple, trivial example
+  }
+
+  <<__EntryPoint>>
+  async function call_hello_world(): Awaitable<void> {
+
+  $aw = hello_world();
+  $result = await $aw; // Suspends until $aw completes
```
Thinking about how we describe async, I suppose the important thing is that you can wait for multiple things at once.

I can't see a good way of writing a small self-contained example for this. How about something like this?

```hack
<<__EntryPoint>>
async function my_example(): Awaitable<void> {
  $user_ids = vec[1, 2, 3];

  // This will initiate all the database requests together, so we
  // spend less time waiting.
  $user_names = await Vec\map_async(
    $user_ids,
    async ($id) ==> await fetch_user_name($id),
  );
  // Execution continues once all the requests have completed.

  echo Str\join($user_names, ", ");
}

async function fetch_user_name(int $_): Awaitable<string> {
  // In practice this would be a database request.
  return "";
}
```

We could even drop the `fetch_user_name` function if you think it's distracting.

AndrewDiMola replied:
Had the same thought here -- Just couldn't find a way to get it all neatly self-contained so I used a trivial example.

Wilfred reviewed (approved):
Some additional suggestions for you, but feel free to land :)

AndrewDiMola committed:
Update website/pages/en/index.js

Co-authored-by: Wilfred Hughes <me@wilfred.me.uk>

AndrewDiMola committed:
Reformat Async example

AndrewDiMola committed:
Merge branch 'landing-page-updates' of https://github.com/AndrewDiMola/hacklang.org into landing-page-updates
